specs/withdrawals.md

```diff
@@ -0,0 +1,181 @@
+   performs the remainder of the verification process internally.
+3. If verification fails, the call reverts. Otherwise the call is forwarded, and the hash is recorded to prevent it from
+   from being replayed.
+
+## The L2 Withdrawer Contract
+
+A withdrawal is initiated by calling the Withdrawer contract's `initiateWithdrawal` function.
+The Withdrawer is a simple predeploy contract at `0x4200000000000000000000000000000000000015` which stores messages
+to be withdrawn.
+
+```js
+interface Withdrawer {
+
+    event WithdrawalMessage(
+        uint256 indexed nonce, // this is a global nonce value for all withdrawal messages
+        address indexed sender,
+        address indexed target,
+        uint256 value,
+        uint256 gasLimit,
+        bytes message
+    );
+
+    function initiateWithdrawal(
+        address target,
+        uint256 gasLimit,
+        bytes data,
+    ) external payable;
+
+    function burn();
+}
+```
+
+If called by a contract, the `initiateWithdrawal` function will alias the `sender` address in the reverse of the
+manner described in the [deposits spec](./deposits.md#address-aliasing).
+
+## The Optimism Portal Contract
+
+The Optimism Portal serves as both the entry and exit point to the Optimism L2. It is a contract which inherits from
+the [DepositFeed](./deposits.md#deposit-contract) contract, and in addition provides the following interface for
+withdrawals:
+
+```js
+interface OptimismPortal {
+
+    event WithdrawalFinalized(bytes32 indexed);
+
+    function finalizeWithdrawalTransaction(
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _gasLimit,
+        bytes calldata _data,
+        uint256 _timestamp,
+        WithdrawalVerifier.OutputRootProof calldata _outputRootProof,
+        bytes calldata _withdrawalProof
+    )
+}
+```
+
+## Withdrawal Verification and Finalization
+
+The following inputs are required to verify and finalize a withdrawal:
+
+- Withdrawal transaction data:
+  - `nonce`: Nonce for the provided message.
+  - `sender`: Message sender address on L2.
+  - `target`: Target address on L1.
+  - `data`: Data to send to the target.
+  - `gasLimit`: Gas to be forwarded to the target.
+- Proof and verification data:
+  - `timestamp`: The L2 timestamp corresponding with the output root.
+  - `outputRootProof`: The 4 values which are used to derive the output root.
+  - `withdrawalProof`: An inclusion proof for the given withdrawal in the withdrawer contract.
+
+These inputs must satisfy the following conditions:
+
+1. The `timestamp` is at least `FINALIZATION_PERIOD` seconds old.
+1. `OutputOracle.l2Outputs(timestamp)` returns a non-zero value `l2Output`.
+2. The `l2Output` value is equal to the hash of the components of the `outputRootProof`
+1. The `withdrawalProof` is a valid inclusion proof demonstrating that a hash of the Withdrawal transaction data
+   is contained in the storage of the Withdrawer contract on L2.
```

Comment:
This is missing a validity condition regarding `outputRootProof`.

Reply:
> This is missing a validity condition regarding `outputRootProof`.

I think it's covered by line 143, how about if I rewrite it to emphasize the proof rather than the output?

```diff
- 1. The `l2Output` value is equal to the hash of the components of the `outputRootProof`
+ 1. The keccak256 hash of the `outputRootProof` values is equal to the `l2Output`.
```

Reply:
Oh sorry, you're right. I think it's fine as-is, so up to you :)